RFCs/0000-pdk-template-v2.md

```diff
@@ -0,0 +1,296 @@
+
+* Template users should be to easily test that their custom template settings work as they expect.
+
+**Template Authors**
+
+* Template authors should be able to easily add/modify templates with clear and documented standards e.g. file locations and good practices
+
+* Template authors should be able to easily generate document from their templates instead of manually keeping the code and documentation in sync.
+
+* Template authors should be able to easily test their templates work as expected, both manually and via automation, like rspec.
+
+* Template authors should be able to advertise what their templates do
+
+* Template authors should be able to use ERB for simple templates, but also put more advanced code into Ruby
+
+### Backwards compatibility
+
+The proposed changes must either be backwards compatible, or be segregated so that Template Users can progressively opt in to version 2.0
+
+## Proposed Implementation Design
+
+### Opt-in instead of opt-out
+
+Currently the templating will apply everything, and Template Users must opt-out.  The new templating system will use a composition model, that is, Template User must opt-in to things.
+
+### Easier to discover, and validate, template settings
+
+Currently Template User need to read the README to determine what settings are available or even valid. This also assumes the README is even correct.  The new templating system will have a metadata, and optionally, a validation layer so that users can discover what templates are available to opt-in to, and optionally, to validate template settings.
+
+### Easier to manage template file locations
+
+The new templating design will have descriptive and well documented file locations
+
+
+**V1 locations**
+
+| Path                      | Description |
+| ------------------------- | ----------- |
+| `/moduleroot`             | Templates get deployed on new module, convert and update |
+| `/moduleroot_init`        | Templates get only deployed when the target module does not yet exist |
+| `/object_templates/*.erb` | Template location for `pdk new ...` generation |
+| `/robocop`                | Custom ruby code to assist in generating rubocop profiles |
+
+**V2 locations**
+
+Where `name` is a unique name for the template
+
+> The `name` should use snake case (lowercase with underscore) to be cross platform compliant. The more descriptive name can be specified in the template.yml file
+
+| Path                               | Description |
+| ---------------------------------- | ----------- |
+| `/<name>/README.md`            (1) | A plain text description of the template. Though not really necessary as most of this would be contained in the `template.yml` file. It is expected that this file will be auto-generated |
+| `/<name>/template.yml`             | YAML file describing the template |
+| `/<name>/template_schema.json` (1) | JSON Schema file describing the valid settings for this template |
```

Comment:
> do we need the full power of json schema here

No, we don't need the full power of JSON schema (e.g. object references) however, we don't have to use all of the schema specification. The example JSON schema below shows a simple, but common, example of an array of strings.

Also, we could add a helper (gem, rake task, whatever) to introspect the `template.yml` default settings and then spit out some boilerplate JSON Schema for the user to fill out.

> mixing yaml for the description and json for the schema adds a lot of complexity.

I agree that have two data types increases cognitive load, but I would disagree with "a lot" of complexity.  The path I've taken here is that YAML should be used where possible as it is more "human-readable" and supports comments.  The JSON is used where the data needs to also be more "machine-readable"

I would argue that using the Bolt metadata or Resource API Style schema is far more complex as it uses Puppet Data Types which is very different to YAML.  Whereas JSON Schema and YAML are far more similar.

> or can we get by with either the bolt task metadata definition or a Resource API style schema? (notably both use puppet data types, and only allow a single layer of top-level keys)

There are many libraries in many languages to support JSON Schema validation and generation, compared to Puppet Data Types.  Using a common validation schema means it will be easier for other people, using other tools, to validate their template settings.  For example; What if we had a "PDK Template Validator" like https://validate.puppet.com/ which could validate template settings client side.